### PR TITLE
sanitycheck: Add option to use gcovr for coverage

### LINKED ISSUE #17626: Change sanitycheck to use 'gcovr' instead of 'lcov'
LCOV is written in Perl and introduces some hardships for people trying to produce coverage reports on Windows. It is also a third-party package that has to be installed, with 1.14 being the minimum version.

gcovr, on the other hand, is written in Python, and is installable with pip; indeed it is already in our scripts/requirements.txt.

Change sanitycheck to use gcovr instead of lcov to produce our code coverage reports.

### PR DESCRIPTION
**Pull request by chrta**
gcovr is already a dependency in scripts/requirements.txt. The visualization
is different, but the functionality should be the same. Tested with gcovr 4.2.

Relates to #17626.

Signed-off-by: Christian Taedcke <hacking@taedcke.com>

### CODE REVIEW

**Comment by chrta** [comment]
When executing, i get a few warnings :( :
~~~
WARNING) GCOV produced the following errors processing /projects/zephyrproject/zephyr/sanity-out/native_posix/samples/application_development/external_lib/sample.app_dev.external_lib/mylib/obj/mylib.gcda:
	Cannot open source file src/mylib.c

	Cannot open source file src/mylib.c

....
	Could not open output file 'src#mylib.c.gcov'

	(gcovr could not infer a working directory that resolved it.)
~~~

**Comment by chrta** [comment]
Another note: Gcovr seems to be a lot slower:

master branch (lcov)
~~~
$ time scripts/sanitycheck -p native_posix -C
...
real	16m49,274s
user	37m59,643s
sys	11m42,091s
~~~

this branch (gcovr)
~~~
real	30m55,836s
user	40m55,605s
sys	12m17,708s
~~~

**Comment by chrta** [comment]
@nashif How about i change this, so the coverage tool is selectable by command line argument? By default lcov is used, optionally gcovr can be selected?

**Comment by nashif** [comment]
> @nashif How about i change this, so the coverage tool is selectable by command line argument? By default lcov is used, optionally gcovr can be selected?

yes, that works.

**Comment by chrta** [comment]
There is a new option now: '--coverage-tool {lcov, gcovr}', by default lcov is used.

File: scripts/sanitycheck [hunk 1]
```diff
@@ -3486,60 +3489,142 @@ def create_gcda_files(extracted_coverage_info):
...
+    def factory(tool):
+        if tool == 'lcov':
+            return Lcov()
+        if tool == 'gcovr':
+            return Gcovr()
+        error("Unsupported coverage tool specified: {}".format(tool))
+
+    def generate(self, outdir):
+        for filename in glob.glob("%s/**/handler.log" % outdir, recursive=True):
+            gcov_data = retrieve_gcov_data(filename)
```
Comment (nashif) [inline scripts/sanitycheck]: can we move this function into the class as well, maybe as a static method, this way we will have all things related to coverage under one class.

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/sanitycheck [hunk 2]
```diff
@@ -3486,60 +3489,142 @@ def create_gcda_files(extracted_coverage_info):
...
+            capture_complete = gcov_data['complete']
+            extracted_coverage_info = gcov_data['data']
+            if capture_complete:
+                create_gcda_files(extracted_coverage_info)
```
Comment (nashif) [inline scripts/sanitycheck]: same thing for this one.

File: scripts/sanitycheck (hunk 1 above)
Comment (chrta) [reply scripts/sanitycheck]: done

File: scripts/sanitycheck (hunk 2 above)
Comment (chrta) [reply scripts/sanitycheck]: done

**Review by nashif — APPROVED** [review]

**Pull request merged**